Automattic/wp-calypso
# composite-checkout: Inject steps as arguments

#### Changes proposed in this Pull Request

A great deal of the `Checkout` form is now provided by customization slots, including the first step (`OrderSummary`), the contact form (`ContactSlot`), and the last step (`ReviewContent`). This is appropriate, but it does highlight an evolving need: the host page wants to control the contents of the form more directly. Using slots means that we have to expect every situation that the host page might want to do and add a slot for it. There is nothing wrong with this approach but it restricts what can be done without changing the checkout component itself.

This PR modifies the `Checkout` component so that instead of taking a series of components that it injects into slots in the steps, it takes an array of _steps_ to display. 

Each _step_ is an object with a particular schema that includes both components and meta-data about the step's behavior. This is similar to how payment methods work now. And just like payment methods, we can provide some steps as constructor functions from within the package; notably the Payment Methods step is likely to be required for every form. The form then connects those steps with "continue" buttons and displays each step when appropriate.

This also helps to address a different issue: it was never very clear what props were being injected into a slotted component and how they were supposed to be used. `summary`, `isActive`, `isComplete` were all used in various combinations that the components had to understand and render correctly. Since each step now has a full object, each version of the step can be a separate component with a clearly defined role (or the same component with different props, but that is up to the author of that component and isn't baked into the system as it was with the slots). The roles represent the original three states of a step and use the properties `activeStepContent`, `incompleteStepContent`, and `completeStepContent`.

The step object also can reference various functions. Most importantly, the `isCompleteCallback` property creates a defined place to perform validation for a step in order to determine if that step is complete.

Finally, some steps are actually just informational blocks and should never be active (eg: the `OrderSummary` slot in the current form). Giving these steps a number is awkward and could cause problems if we ever want one between other numbered steps. With this change, these steps can set the property `hasStepNumber` to false, and they will not be counted in the step numbering system nor will it be possible to make them active.

Here's an example of a step object:

```js
{
	id: 'payment-method',
	className: 'checkout__payment-methods-step',
	hasStepNumber: true,
	titleContent: <CheckoutPaymentMethodsTitle />,
	activeStepContent: <CheckoutPaymentMethods isComplete={ false } />,
	incompleteStepContent: null,
	completeStepContent: <CheckoutPaymentMethods summary isComplete={ true } />,
	isCompleteCallback: ( { paymentData } ) => {
		const { billing = {} } = paymentData;
		if ( ! billing.country ) {
			return false;
		}
		return true;
	},
	isEditableCallback: ( { paymentData } ) => {
		return ( paymentData.billing ) ? true : false;
	},
	getEditButtonAriaLabel: () => translate( 'Edit the payment method' ),
	getNextStepButtonAriaLabel: () => translate( 'Continue with the selected payment method' ),
},
```

#### Testing instructions

- Add a file called `packages/composite-checkout/demo/private.js` and inside it put the following: `export const stripeKey = 'TEST KEY HERE'`;
- Run `npm run composite-checkout-demo`.
- Verify that the form works as expected.
- Run calypso with the `composite-checkout-wpcom` flag and verify that the form works as expected there as well.

Note that this PR adds a bunch of tests to verify the behavior of steps. This is actually another big benefit to this strategy; it gives us the ability to more easily write tests for the behaviors of various steps.
